From: Thorsten Glaser <tg@mirbsd.org>
Subject: change location of the socket path to something less racy
 also, rely on the initscript to create it with correct permissions and ownership
--- a/common/os_calls.c
+++ b/common/os_calls.c
@@ -113,17 +113,9 @@ g_mk_temp_dir(const char *app_name)
 {
     if (!g_directory_exist(XRDP_SOCKET_PATH))
     {
-        if (!g_create_dir(XRDP_SOCKET_PATH))
-        {
-            /* if failed, still check if it got created by someone else */
-            if (!g_directory_exist(XRDP_SOCKET_PATH))
-            {
                 printf("g_mk_temp_dir: g_create_dir(%s) failed\n",
                        XRDP_SOCKET_PATH);
                 return 1;
-            }
-        }
-        g_chmod_hex(XRDP_SOCKET_PATH, 0x3777);
     }
     return 0;
 }
--- a/instfiles/xrdp.service
+++ b/instfiles/xrdp.service
@@ -9,8 +9,11 @@ Type=forking
 PIDFile=/var/run/xrdp.pid
 EnvironmentFile=-/etc/sysconfig/xrdp
 EnvironmentFile=-/etc/default/xrdp
+PermissionsStartOnly=true
+ExecStartPre=/bin/sh /usr/share/xrdp/socksetup
 ExecStart=/usr/sbin/xrdp $XRDP_OPTIONS
 ExecStop=/usr/sbin/xrdp $XRDP_OPTIONS --kill
+ExecStopPost=/bin/rm -rf /var/run/xrdp/sockdir
 
 [Install]
 WantedBy=multi-user.target
--- a/xorg/X11R7.6/rdp/rdpmisc.c
+++ b/xorg/X11R7.6/rdp/rdpmisc.c
@@ -527,7 +527,7 @@ g_socket_dir(void)
     socket_dir = getenv("XRDP_SOCKET_PATH");
     if (socket_dir == NULL || socket_dir[0] == '\0')
     {
-        socket_dir = "/tmp/.xrdp";
+        socket_dir = "/var/run/xrdp/sockdir";
     }
 
     return socket_dir;
--- a/xorg/X11R7.6/rdp/rdpup.c
+++ b/xorg/X11R7.6/rdp/rdpup.c
@@ -1239,13 +1239,8 @@ rdpup_init(void)
 
     if (!g_directory_exist(socket_dir))
     {
-        if (!g_create_dir(socket_dir))
-        {
             LLOGLN(0, ("rdpup_init: g_create_dir(%s) failed", socket_dir));
             return 0;
-        }
-
-        g_chmod_hex(socket_dir, 0x3777);
     }
 
     i = atoi(display);
--- a/xorgxrdp/module/rdpClientCon.c
+++ b/xorgxrdp/module/rdpClientCon.c
@@ -1178,15 +1178,8 @@ rdpClientConInit(rdpPtr dev)
     socket_dir = g_socket_dir();
     if (!g_directory_exist(socket_dir))
     {
-        if (!g_create_dir(socket_dir))
-        {
-            if (!g_directory_exist(socket_dir))
-            {
                 LLOGLN(0, ("rdpup_init: g_create_dir(%s) failed", socket_dir));
                 return 0;
-            }
-        }
-        g_chmod_hex(socket_dir, 0x3777);
     }
     i = atoi(display);
     if (i < 1)
--- a/xorgxrdp/module/rdpMisc.c
+++ b/xorgxrdp/module/rdpMisc.c
@@ -388,7 +388,7 @@ g_socket_dir(void)
     socket_dir = getenv("XRDP_SOCKET_PATH");
     if (socket_dir == NULL || socket_dir[0] == '\0')
     {
-        socket_dir = "/tmp/.xrdp";
+        socket_dir = "/var/run/xrdp/sockdir";
     }
 
     return socket_dir;
--- a/xrdp/xrdp.ini
+++ b/xrdp/xrdp.ini
@@ -138,10 +138,10 @@ xrdpvr=true
 tcutils=true
 
 ; for debugging xrdp, in section xrdp1, change port=-1 to this:
-#port=/tmp/.xrdp/xrdp_display_10
+#port=/var/run/xrdp/sockdir/xrdp_display_10
 
 ; for debugging xrdp, add following line to section xrdp1
-#chansrvport=/tmp/.xrdp/xrdp_chansrv_socket_7210
+#chansrvport=/var/run/xrdp/sockdir/xrdp_chansrv_socket_7210
 
 
 ;
